lib/semian/lru_hash.rb

```diff
@@ -0,0 +1,53 @@
+class LRUHash
+  extend Forwardable
+  def_delegators :@table, :size, :count, :empty?, :keys, :clear, :delete
+
+  attr_reader :table, :minimum_time_in_lru
+  def initialize(minimum_time_in_lru = 300)
+    @minimum_time_in_lru = minimum_time_in_lru
+    @table = {}
+    @lock = Mutex.new
+  end
+
+  def set(key, value)
+    @lock.synchronize {
+      @table.delete(key)
+      @table[key] = value
+      clear_unused_resources
+    }
+    value
+  end
+
+  def get(key)
+    @lock.synchronize {
+      if found = @table.delete(key)
+        @table[key] = found
+      end
+    }
+    @table[key]
+  end
+
+  def []=(key, value)
+    set(key, value)
+  end
+
+  def [](key)
+    get(key)
+  end
+
+  private
+
+  def clear_unused_resources
+    cleaned_elements = 0
+    @table.each do |key, value|
+      break if cleaned_elements >= 2
```

Comment:
From the benchmarks, we can see that it does not bring much value to remove only 2 resources per set. I think we should be fine with removing all unused resources and not limiting them.

Reply:
Note that your benchmarks are going to be fairly contrived. This may actually work well in the real world, but I have no idea, and no proof—in the absence of that, I think you should remove this complexity because it's counter-intuitive. Counterintuitive behaviour is OK in the presence of a proof, but otherwise, 🔥

Reply:
I must say I disagree here. I don't have proof but I'd expect that iterating over `1000` items hash is slower than iterating over a `2` items. You might say that the max size of resources in core is `100` so it's really a comparison between `2` and `100` but I expect the resources count to increase when we use this for things like webhooks.

Assuming we go with the `updated_at` approach rather than `created_at`, I think this can be solved via `@table.take(2).each do |key, value|` which is hardly much more complexity.